### ISSUE TITLE
Fix S2583: Rule should not report false positives in finally block

### ISSUE THREAD

**Issue by Duetting** [SonarSource/sonar-csharp#639]
Rule 2583 is even shown in finally blocks, e. g. if a disposable object needs to be returned from a method and the code from [FxCop CA2000](https://msdn.microsoft.com/library/ms182289.aspx) is used:

```C#
public SerialPort OpenPort2(string portName)
{
   SerialPort tempPort = null;
   SerialPort port = null;
   try
   {
      tempPort = new SerialPort(portName);
      tempPort.Open();
      SomeMethod();
      //Add any other methods above this line
      port = tempPort;
      tempPort = null;
      
   }
   finally
   {
//S2583 is shown here for tempPort, which is reachable if an exception is thrown, e. g. in SomeMethod
      if (tempPort != null)
      {
         tempPort.Close();
      }
   }
   return port;
}
```

SonarLint Version 3.3.0.1663
Visual Studio Community 2015 Update 3

**Comment by Duetting** [comment]
I've just seen that this could be a possible duplicate of #621.

**Comment by Evangelink** [comment]
Hi @Duetting this is actually linked to #617.

**Comment by Evangelink** [comment]
@SonarSource/dotnet We could try to improve our CFG builder by either:
1/ Detecting variables checked against null in finally block and create a new branch to the finally/catch after their initialization in the try
2/ We could go for a naive approach and assume the initialization is going to happen on the first line of the try and so add a path to finally/catch after the first line (executable).  Of course if multiple variables are initialized or if the initialization doesn't occur on first line we will have the same issue but this should cover most of the cases.
